exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -493,11 +477,36 @@ public Builder withTimeService(TimeProvider timeProvider) {
     public TestKit build() {
       checkCorrectServiceNumber(services.size());
       checkCorrectValidatorNumber();
-      return newInstance(services, nodeType, validatorCount, timeProvider);
+      List<TestKitServiceInstances> testKitServiceInstances = mergeServiceSpecs();
+      return newInstance(testKitServiceInstances, nodeType, validatorCount, timeServiceSpec);
+    }
+
+    /**
+     * Turn collection of service instances into a list of
+     * {@linkplain TestKitServiceInstances} objects for native to work with.
+     */
+    private List<TestKitServiceInstances> mergeServiceSpecs() {
+      Set<ServiceArtifactId> serviceArtifactIds = services.keySet();
+      checkArgument(serviceArtifactIds.containsAll(serviceArtifactFilenames.keySet()),
+          "All service instances that are deployed should also be instantiated"
+              + " and vice versa.");
+      return serviceArtifactIds.stream()
+          .map(this::aggregateServiceSpecs)
+          .collect(toList());
+    }
+
+    /**
+     * Aggregates service instances specifications of a given service artifact id as a
+     * {@linkplain TestKitServiceInstances} object.
+     */
+    private TestKitServiceInstances aggregateServiceSpecs(ServiceArtifactId artifactId) {
+      String artifactFilename = serviceArtifactFilenames.get(artifactId);
```

Comment:
The client code streams from the keys of this map (serviceArtifactFilenames), and as its  next step, adds the values from the same map. Why doesn't it stream entries then?